### PR TITLE
[Refactor] R3Det and S2A-Net

### PR DESCRIPTION
**Pull request by zytx121**
Add `RefineSingleStageDetector` to support R3Det, Cascade R3Det, S2A-Net and Cascade S2A-Net.
Refer to https://github.com/open-mmlab/mmrotate/pull/294
- [x] Align test
- [x] Align train
- [x] UT

## Experiments
s2anet_r50_fpn_1x_dota_le135.py
report AP50: 73.91
AP50: 0.7372947528583106
AP75: 0.3631685947094186
mAP: 0.4015696604488543

cascade_s2anet_r50_fpn_1x_dota_le135.py
AP50: 0.7390151597639424
AP75: 0.386362895638701
mAP: 0.40812074188598924

r3det_r50_fpn_1x_dota_oc.py
report AP50: 69.80
AP50: 0.6914788143725653
AP75: 0.38191651495229434
mAP: 0.38352609799438936

r3det_tiny_r50_fpn_1x_dota_oc
report AP50: 70.18
AP50: 0.6983682393653567
AP75: 0.35542363646946357
mAP: 0.3773277534440793

### CODE REVIEW

File: mmrotate/models/utils/align.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    """DCN Align Module.
+
+    All args are from DeformConv2dPack.
+    TODO: maybe use build_conv_layer is more flexible.
+    """
+
+    def forward(self, x: List[Tensor],
+                anchors: List[List[Tensor]]) -> List[Tensor]:
+        """Forward function."""
+        return super(DCNAlignModule, self).forward(x)
```
Comment (liuyanyi) [inline mmrotate/models/utils/align.py:121]: The x here is a list of Tensor, here should be:
```
        return [super(DCNAlignModule, self).forward(xi) for xi in x]
```

File: mmrotate/models/utils/align.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            convolution layer. Defaults to None.
+        norm_cfg (:obj:`ConfigDict` or dict, optional): Config dict for
+            normalization layer. Defaults to None.
+    """
+
+    def __init__(self,
+                 feat_channels: int,
+                 strides: List[int],
+                 conv_cfg: OptConfigType = None,
+                 norm_cfg: OptConfigType = None) -> None:
```
Comment (liuyanyi) [inline mmrotate/models/utils/align.py:141]: is 'conv_cfg' and 'norm_cfg' useful here? maybe it canbe remove or use `ConvModule` in `init_layers`?

File: mmrotate/models/detectors/refine_single_stage.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        backbone (:obj:`ConfigDict` or dict): The backbone module.
+        neck (:obj:`ConfigDict` or dict): The neck module.
+        bbox_head_init (:obj:`ConfigDict` or dict): The bbox head module of
+            the first stage.
+        bbox_head_refine (list[:obj:`ConfigDict` | dict]): The bbox head
+            module of the refine stage.
+        train_cfg (:obj:`ConfigDict` or dict, optional): The training config
+            of ATSS. Defaults to None.
+        test_cfg (:obj:`ConfigDict` or dict, optional): The testing config
+            of ATSS. Defaults to None.
```
Comment (liuyanyi) [inline mmrotate/models/detectors/refine_single_stage.py:29]: ATSS -> RefineSingleStageDetector

**Review by liuyanyi — APPROVED** [review]
LGTM

File: mmrotate/models/dense_heads/r3_head.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+            device (torch.device | str): Device for returned tensors.
+                Defaults to cuda.
+
+        Returns:
+            tuple:
+
+                - anchor_list (list[list[Tensor]]): Anchors of each image.
+                - valid_flag_list (list[list[Tensor]]): Valid flags of each
+                  image.
+        """
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/r3_head.py:165]: docstring format

File: mmrotate/models/dense_heads/r3_head.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+            list[:obj:`InstanceData`]: Object detection results of each image
+            after the post process. Each item usually contains following keys.
+
+                - scores (Tensor): Classification scores, has a shape
+                  (num_instance, )
+                - labels (Tensor): Labels of bboxes, has a shape
+                  (num_instances, ).
+                - bboxes (Tensor): Has a shape (num_instances, 4),
+                  the last dimension 4 arrange as (x1, y1, x2, y2).
+        """
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/r3_head.py:227]: docstring format

File: mmrotate/models/dense_heads/s2a_head.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+        Args:
+            x (Tensor): Features of a single scale level.
+
+        Returns:
+            tuple:
+                cls_score (Tensor): Cls scores for a single scale level
+                    the channels number is num_anchors * num_classes.
+                bbox_pred (Tensor): Box energies / deltas for a single scale
+                    level, the channels number is num_anchors * 4.
```
Comment (RangiLyu) [inline mmrotate/models/dense_heads/s2a_head.py:139]: format

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**